SaraAlert/SaraAlert
# SARAALERT-394: Admin Audit capability

# Description
Jira Ticket: SARAALERT-394

Audit logs available for each visible user through the Admin table.

Branched off of [this PR](https://github.com/SaraAlert/SaraAlert/pull/505).

# Demo/Screenshots
View from Admin Table:

![image](https://user-images.githubusercontent.com/42656458/101218245-be203f00-3650-11eb-8e9d-ba37c063f407.png)

Example view of a specific user's Audit Log:

![image](https://user-images.githubusercontent.com/42656458/101218297-cd9f8800-3650-11eb-89a2-fd20e16fc570.png)


# Important Changes

`app/controllers/jurisdictions_controller.rb`
- New method to get all jurisdiction paths

`app/controllers/users_controller.rb`
- Method to get all audits associated with a given user

`app/javascript/components/admin/AdminTable.js`
- New column for the admin to access a specific user's audit log

`app/javascript/components/admin/AuditModal.js`
- Modal for displaying a given user's audits in a table. 
- Leverages the CustomTable layout

`app/javascript/components/layout/CustomTable.js`
- Adjustment to allow a table to have auditable records
- Adjustment for when a table shouldn't have select-able rows (taken from this [Draft PR](https://github.com/SaraAlert/SaraAlert/pull/531))

`app/models/user.rb`
- Permissions set for which users can view audit logs
- Specifies which data elements associated with a user will be used for the audit log entries

`config/routes.rb`
- New route to get all jurisdiction paths

`db/schema.rb`
- New Audits table

# Testing
This fix was tested on the following browsers (submitter must check all those that apply):

While logged in as an admin or super user, perform any of the following actions on a specific user:
- Lock/Unlock the account
- Enable/Disable API access
- Change their role
- Change their jurisdiction
- Change their email
- Reset their email
- (2FA reset - will only work if the user has 2FA access required)
* [ ] For that user, click on the icon in the Audits column & verify a modal appears with a table of audits
* [ ] In the table, ensure there are audits visible for the various actions that were performed on that user's account.
* [ ] Test that sorting works properly for the "Triggered By" and "Timestamp" columns
* [ ] Pagination should default to displaying 25 entries. Test changing the number of displayed entries and ensure the table updates properly.
